File: lwaf_3D/IRotated.java
package lwaf_3D;

import lwaf_core.vec3;

@SuppressWarnings("UnusedReturnValue")
public interface IRotated<R> {
    vec3 getRotation();
    R setRotation(vec3 rotation);

    default R rotateBy(vec3 rotation) {
        return setRotation(getRotation().add(rotation));
    }

    default R setRotation(float x, float y, float z) {
        return setRotation(new vec3(x, y, z));
    }

    default R rotateBy(float x, float y, float z) {
        return setRotation(getRotation().add(new vec3(x, y, z)));
    }
}

File: lwaf_3D/IRotated.kt
package lwaf_3D

import lwaf_core.plus
import lwaf_core.vec3

interface IRotated<R> {
    val rotation: vec3
    fun setRotation(rotation: vec3): R

    fun rotateBy(rotation: vec3): R {
        return setRotation(this.rotation + rotation)
    }

    fun setRotation(x: Float, y: Float, z: Float): R {
        return setRotation(vec3(x, y, z))
    }

    fun rotateBy(x: Float, y: Float, z: Float): R {
        return setRotation(rotation + vec3(x, y, z))
    }
}
